### PR TITLE
Make CI skip check more robust

### PR DESCRIPTION
**Pull request by mandeepsandhu**
Implements the 2 TODOs which make the CI check robust to erroneous
skips.

The script will now:
- not skip CI if previous CI was not successful
- will check all files since the last successful CI run for determining
  if the commits can skip CI

### CODE REVIEW

File: scripts/needs_ci_run
Comment (mandeepsandhu) [inline scripts/needs_ci_run:10]: Oos, I forgot to remove this! I'll do it in the next push

File: .circleci/config.yml
Comment (mandeepsandhu) [inline .circleci/config.yml:127]: Moved this back to its original place, since I needed to install python `requests` package and thought it'd be better to keep all installs at 1 place. This does add ~2+ mins before the skip CI script can run. But I guess that shouldn't be a big deal.

**Comment by mandeepsandhu** [comment]
@kevinkreiser I think all the drawbacks of the previous attempt at determining if a change can skip CI have been addressed. Let me know if there are issues you see with this change.

CC: @PureW @nilsnolde

**Comment by mandeepsandhu** [comment]
All tests were run on this branch: https://github.com/valhalla/valhalla/pull/3031

File: scripts/needs_ci_run
Comment (mandeepsandhu) [inline scripts/needs_ci_run:53]: [circle API doc](https://circleci.com/docs/api/#recent-builds-for-a-single-project)

**Comment by nilsnolde** [comment]
What do you think integrating the „skip for draft PR“ for circle CI here? Ref https://github.com/valhalla/valhalla/pull/3014#issuecomment-821970848

**Comment by mandeepsandhu** [comment]
> What do you think integrating the „skip for draft PR“ for circle CI here? Ref [#3014 (comment)](https://github.com/valhalla/valhalla/pull/3014#issuecomment-821970848)

Definitely a good idea to consider, although I'm not sure if everyone would want to skip CI for draft PRs (eg I've relied on CI for testing some of my changes on macos). I'm thinking this should be something that a user explicitly asks for. Which reminds me of the circle CIs `[skip ci]` tag that will cause it to not build such commits. Maybe users can use that instead?

**Comment by nilsnolde** [comment]
It's a pretty subjective topic I guess. As for me I really like not to waste so many CI hours and I'm really bad at putting `[ci skip]` ;) I usually don't need/want CI before final review and if I do on rare occasions I find it ok to un-draft, do a small commit and let CI run, then draft again. It'd be best if CI ran automatically when I un-draft without committing again, but we only get that on platforms who support this draft PR stuff natively.

But I'm just a tiny speck here;) maybe someone else has an opinion on it.

**Review by nilsnolde — APPROVED** [review]

**Comment by kevinkreiser** [comment]
lgtm

**Review by kevinkreiser — APPROVED** [review]

**Review by merkispavel — DISMISSED** [review]
re-approve

**Comment by mandeepsandhu** [comment]
> re-approve

Thanks. I've put the fix for osx in https://github.com/valhalla/valhalla/pull/3037. I'll update this PR once thats merged.

**Review by merkispavel — APPROVED** [review]
re-approve

**Pull request merged**
